rllib/examples/custom_input_api.py

```diff
@@ -0,0 +1,127 @@
+"""Example of creating a custom input api
+
+Custom input apis are useful when your data source is in a custom format or
+when it is necessary to use an external data loading mechanism.
+In this example, we train an rl agent on user specified input data.
+Instead of using the built in JsonReader, we will create our own custom input
+api, and show how to pass config arguments to it.
+
+To train CQL on the pendulum environment:
+$ python custom_input_api.py --input-files=../tests/data/pendulum/enormous.zip
+"""
+
+import argparse
+import os
+from pathlib import Path
+
+import ray
+from ray import tune
+from ray.rllib.offline import JsonReader, ShuffledInput, IOContext, InputReader
+from ray.tune.registry import register_input
+
+parser = argparse.ArgumentParser()
+parser.add_argument(
+    "--run",
+    type=str,
+    default="CQL",
+    help="The RLlib-registered algorithm to use.")
+parser.add_argument(
+    "--framework",
+    choices=["tf", "tf2", "tfe", "torch"],
+    default="tf",
+    help="The DL framework specifier.")
+parser.add_argument("--stop-iters", type=int, default=100)
+parser.add_argument(
+    "--input-files",
+    type=str,
+    default=os.path.join(
+        os.path.dirname(os.path.abspath(__file__)),
+        "../tests/data/pendulum/small.json"))
+
+
+class CustomJsonReader(JsonReader):
+    """
+    Example custom InputReader implementation (extended from JsonReader).
+
+    This gets wrapped in ShuffledInput to comply with offline rl algorithms.
+    """
+
+    def __init__(self, ioctx: IOContext):
+        """
+        The constructor must take an IOContext to be used in the input config.
+        Args:
+            ioctx (IOContext): use this to access the `input_config` arguments.
+        """
+        super().__init__(ioctx.input_config["input_files"], ioctx)
+
+
+def input_creator(ioctx: IOContext) -> InputReader:
+    """
+    The input creator method can be used in the input registry or set as the
+    config["input"] parameter.
+
+    Args:
+        ioctx (IOContext): use this to access the `input_config` arguments.
+
+    Returns:
+        instance of ShuffledInput to work with some offline rl algorithms
+    """
+    return ShuffledInput(CustomJsonReader(ioctx))
+
+
+if __name__ == "__main__":
+    ray.init()
+    args = parser.parse_args()
+
+    # Bazel makes it hard to find files specified in `args` (and `data`).
+    # Look for them here.
+    if not os.path.exists(args.input_files):
+        # This script runs in the ray/rllib/examples dir.
+        rllib_dir = Path(__file__).parent.parent
+        input_dir = rllib_dir.absolute().joinpath(args.input_files)
+        args.input_files = str(input_dir)
```

Comment:
Hmm, what's this for? Maybe we can just print the link to json and raise an error if the file doesn't exist?

Reply:
This was present in other examples so I just copied it over. It seems like it is unnecessary though.